## _P. picta_ SEX-DETector

Pipeline to analyse SNP segregation patterns and to identify sex-linked genes from RNA-seq data of a cross (parents and offspring of both sexes) using **[SEX-DETector](https://gitlab.in2p3.fr/sex-det-family/sex-detector/-/tree/master/)**.
See Main/Ppicta_SEXDETector_Scripts for scripts used in this README.

### a. Filter Fastq Files
**[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)**: Quality check your fastq files. Example command:

    fastqc  NS.1547.004.NEBNext_dual_i7_H8---NEBNext_dual_i5_H8.G5_Sire_R2.fastq.gz -o guyana_picta_QC

* Important QC reports include: Per base sequence quality, Per sequence quality scores, Per seq GC content, Overrepresented sequences


**[MultiQC](https://multiqc.info/)**: Merge multiple FastQC files. Ensure that all the fastqcs are in the same directory. Example command:

    multiqc . -o multiqc/

**[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)**: Trim off adaptors. Example command:

    java -jar /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 NS.1547.003.NEBNext_dual_i7_A6---NEBNext_dual_i5_A6.T1_F6_R1.fastq.gz NS.1547.003.NEBNext_dual_i7_A6---NEBNext_dual_i5_A6.T1_F6_R2.fastq.gz T1-F6_output_forward_paired.fq.gz T1-F6_output_forward_unpaired.fq.gz T1-F6_output_reverse_paired.fq.gz T1-F6_output_reverse_unpaired.fq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:50

Can also use the python script to do trimming on multiple files at once. Keep in mind that all the fastq files should contain the uniqid (in this case, "R1/R2"). Example command:
    
    python 01.trimmomatic.py guyana_picta_reads/ /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar /Linux/Trimmomatic-0.36/adapters/TruSeq3-PE.fa >trimming_guyanese_picta.txt

* Run FastQC on the trimmed sequences and compare results.

**[HISAT2](https://daehwankimlab.github.io/hisat2/)**: Map reads to verify sample integrity. Map trimmed reads to a reference genome (Poecilia reticulata) and check the mapping rate for each sample. 

* Build HISAT2 Index: Build index from reference genome

      hisat2-build -f suriname_picta/Guppy_Poecilia_reticulata_refgenome guyana_picta_reads/guppy_reference >hisat2-build.txt
    
* Run HISAT2 Mapping: Example Command:

      hisat2 guppy_reference -1 Trimmed_Reads/T1-F6_A6_R1_003.output_forward_paired.fq.gz -2 Trimmed_Reads/T1-F6_A6_R2_003.output_reverse_paired.fq.gz --phred33 -q -p 12 --no-discordant --no-mixed --no-unal --dta -S ../hisat_alignment --met-file ../hisat_alignment_stats



### b. Constructring _de novo_ assembly

**[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)**: Build _de novo_ assembly. Trinity has a three step-process (Inchworm, Chrysalis, and Butterfly), some of the steps can take a long time. It is recommended to build the assembly in steps.

* Make a tab-delimited file that includes all trimmed forward and reverse sample file names (together with their paths). This will ensure that all samples are used to create the assembly. File should be of the format: 
    sample_name \t sample_forward.fq.gz \t sample_reverse.fq.gz 

Then run:
*  Inchworm step:
*  
      trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_run_chrysalis >trinity_normalization_inchworm.txt
      
*  Chrysalis step:
*  
      trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_distributed_trinity_exec >trinity_chrysalis.txt
      
*  Butterfly step:
*  
      trinity2.11.0/Trinity  --seqType fq --samples_file picta_reads.txt --CPU 32 --max_memory 1500G >trinity_full.txt

* Note the Trinity output format. You will get a '**[Trinity.fasta](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Output-of-Trinity-Assembly)**' output in a new 'trinity_out_dir/' that is created when Trinity is run. Trinity clusters transcripts loosely into 'genes' and contains 'isoforms' of those genes.

Custom script is used to determine assembly statistics (**03.assembly_stats.py**). Script takes the trinity output folder, searches for the assembly fasta file and calculates basic stats.

    python 03.assembly_stats.py trinity_out_dir/



### c. Filtering Transcriptome 
Steps to filter transcriptome, removing redundancy,non-coding RNA and transcripts without an open reading frame

* **[RSEM](https://deweylab.github.io/RSEM/)** & **[Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) Using Trinity Scripts**: Map reads to the _de novo_ transcriptome. 
Trinity has scripts that measure gene expression (RSEM) of reads and aligns them to the _de novo_ assembly (Bowtie2). These scripts can be found in the 'util' directory when Trinity is installed. However, both softwards must be installed and their PATH must be set in order to run the Trinity scripts. Run the following scripts:
* Prepare reference for alignment and abundance estimation:
* Run alignment and abundance esetimation:

Custom scripts are used to filter transcriptome to remove redundancy (**04.get-best-isoform.py** and **05.get-best-isoform-fasta.py**):

* **[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)**: Identify non-coding RNA (ncRNA): Download fasta file of ncRNA from closely related species from **[ENSEMBL](http://uswest.ensembl.org/index.html)** e.g. Oryzias_latipes.MEDAKA1.ncrna.fa. If using BLAST on command line, **[BLAST Commandline Manual](https://www.ncbi.nlm.nih.gov/books/NBK279690/)**.

* Build BLAST index:
* BLAST transcript to ncRNA of related species:

Custom scripts are used to identify top BLAST hit (**06.get-ncrna.py**) and filter the assembly (**07.filter-assembly-ncrna.py**):

* **[Transdecoder](https://github.com/TransDecoder/TransDecoder/wiki)**: Remove transcripts without open reading frames (ORFs).
* Extract long ORFs:
* Final coding region predictions:


### d. SEX-DETector
SEX-DETector requires a gametologs to coassemble into one single transcript. We need to further assemble the transcripts into contigs.

* **[CAP3](http://seq.cs.iastate.edu/cap3.html)**: Assemble contigs.
* Merge CAP3 singlets and contigs:

* **[BWA](http://bio-bwa.sourceforge.net/)**: Map trimmed reads to the final assembly
* Build BWA index:
* Align reads:
* Generate SAM format:

* **[SAMtools](http://www.htslib.org/doc/samtools.html)**: Convert SAMs to BAMs
* Build index:
* Compress into BAM files:
* Order individual BAm files:

* **[reads2snp](https://kimura.univ-montp2.fr/PopPhyl/index.php?section=tools)**: Genotype individuals at each locus. Note the output files:
* .alr - 
* .gen -
* Generate the gen_summary file -
* Run SEX-DETector

