## _P. picta_ SEX-DETector

Pipeline to analyse SNP segregation patterns and to identify sex-linked genes from RNA-seq data of a cross (parents and offspring of both sexes) using **[SEX-DETector](https://gitlab.in2p3.fr/sex-det-family/sex-detector/-/tree/master/)**.
Scripts can be found in [Poecilia_picta_Evol_Hist/Ppicta_SEXDETector_Scripts](https://github.com/ljmfong/Poecilia_picta_Evol_Hist/tree/main/Ppicta_SEXDETector_Scripts). 

### a. Filter Fastq Files
**[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)**: Quality check your fastq files. Example command:

    fastqc  NS.1547.004.NEBNext_dual_i7_H8---NEBNext_dual_i5_H8.G5_Sire_R2.fastq.gz -o guyana_picta_QC

* Important QC reports include: Per base sequence quality, Per sequence quality scores, Per seq GC content, Overrepresented sequences


**[MultiQC](https://multiqc.info/)**: Merge multiple FastQC files. Ensure that all the fastqcs are in the same directory. Example command:

    multiqc . -o multiqc/

**[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)**: Trim off adaptors. Example command:

    java -jar /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 NS.1547.003.NEBNext_dual_i7_A6---NEBNext_dual_i5_A6.T1_F6_R1.fastq.gz NS.1547.003.NEBNext_dual_i7_A6---NEBNext_dual_i5_A6.T1_F6_R2.fastq.gz T1-F6_output_forward_paired.fq.gz T1-F6_output_forward_unpaired.fq.gz T1-F6_output_reverse_paired.fq.gz T1-F6_output_reverse_unpaired.fq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:50

Can also use the python script to do trimming on multiple files at once. Keep in mind that all the fastq files should contain the uniqid (in this case, "R1/R2"). Example command:
    
    python 01.trimmomatic.py guyana_picta_reads/ /Linux/Trimmomatic-0.36/trimmomatic-0.36.jar /Linux/Trimmomatic-0.36/adapters/TruSeq3-PE.fa >trimming_guyanese_picta.txt

* Run FastQC on the trimmed sequences and compare results.

**[HISAT2](https://daehwankimlab.github.io/hisat2/)**: Map reads to verify sample integrity. Map trimmed reads to a reference genome (Poecilia reticulata) and check the mapping rate for each sample. 

* Build HISAT2 Index: Build index from reference genome

      hisat2-build -f suriname_picta/Guppy_Poecilia_reticulata_refgenome guyana_picta_reads/guppy_reference >hisat2-build.txt
    
* Run HISAT2 Mapping: Example Command:

      hisat2 guppy_reference -1 Trimmed_Reads/T1-F6_A6_R1_003.output_forward_paired.fq.gz -2 Trimmed_Reads/T1-F6_A6_R2_003.output_reverse_paired.fq.gz --phred33 -q -p 12 --no-discordant --no-mixed --no-unal --dta -S ../hisat_alignment --met-file ../hisat_alignment_stats



### b. Constructring _de novo_ assembly

**[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)**: Build _de novo_ assembly. Trinity has a three step-process (Inchworm, Chrysalis, and Butterfly), some of the steps can take a long time. It is recommended to build the assembly in steps.

* Make a tab-delimited file that includes all trimmed forward and reverse sample file names (together with their paths). This will ensure that all samples are used to create the assembly. File should be of the format: 
    sample_name \t sample_forward.fq.gz \t sample_reverse.fq.gz 

Then run:
*  Inchworm step:


      trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_run_chrysalis >trinity_normalization_inchworm.txt
      
*  Chrysalis step:


      trinity2.11.0/Trinity --seqType fq --samples_file picta_reads.txt --CPU 32 --normalize_max_read_cov 50 --max_memory 1500G --no_distributed_trinity_exec >trinity_chrysalis.txt
      
*  Butterfly step:


      trinity2.11.0/Trinity  --seqType fq --samples_file picta_reads.txt --CPU 32 --max_memory 1500G >trinity_full.txt

Note the Trinity output format. You will get a '**[Trinity.fasta](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Output-of-Trinity-Assembly)**' output in a new 'trinity_out_dir/' that is created when Trinity is run. Trinity clusters transcripts loosely into 'genes' and contains 'isoforms' of those genes. Custom script is used to determine assembly statistics (**03.assembly_stats.py**). Script takes the trinity output folder, searches for the assembly fasta file and calculates basic stats.

    python 03.assembly_stats.py trinity_out_dir/



### c. Filtering Transcriptome 
Steps to filter transcriptome, removing redundancy,non-coding RNA and transcripts without an open reading frame

**[RSEM](https://deweylab.github.io/RSEM/)** & **[Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) Using Trinity Scripts**: Map reads to the _de novo_ transcriptome. 
Trinity has scripts that measure gene expression (RSEM) of reads and aligns them to the _de novo_ assembly (Bowtie2). These scripts can be found in the 'util' directory when Trinity is installed. However, both softwards must be installed and their PATH must be set in order to run the Trinity scripts. Run the following scripts:
* Prepare reference for alignment and abundance estimation:

        trinity2.11.0/util/align_and_estimate_abundance.pl --transcripts trinity_out_dir/Trinity.fasta --est_method RSEM --aln_method bowtie2 --trinity_mode --prep_reference >prep_rsem_reference.txt
        
* Run alignment and abundance esetimation: Example command:

        trinity2.11.0/util/align_and_estimate_abundance.pl --transcripts trinity_out_dir/Trinity.fasta --seqType fq --left suriname_picta/Trimmed_Reads/T1-Dame_R1_098.output_forward_paired_50bp.fastq.gz --right suriname_picta/Trimmed_Reads/T1-Dame_R2_098.output_reverse_paired_50bp.fastq.gz --est_method RSEM --aln_method bowtie2 --trinity_mode --output_dir picta_rsem/T1-Dame/ --bowtie2_RSEM '--no-mixed --no-discordant --gbar 1000 --end-to-end -k 200 --phred33-quals' >T1-Dame-trinity-rsem.txt 2>T1-Dame-trinity-rsem-alignment.txt


Custom scripts are used to filter transcriptome to remove redundancy (**04.get-best-isoform.py** and **05.get-best-isoform-fasta.py**):

        python 04.get-best-isoform.py picta_rsem/ trinity_out_dir/Trinity.fasta best_isoform.txt >04_best_isoform.txt
        python 05.get-best-isoform-fasta.py trinity_out_dir/Trinity.fasta best_isoform.txt >05_get_best_isoform.txt

It will save in the trinity_out_dir folder as Trinity.bestisoform.fa 

**[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)**: Identify non-coding RNA (ncRNA): Download fasta file of ncRNA from closely related species from **[ENSEMBL](http://uswest.ensembl.org/index.html)** e.g. Oryzias_latipes.MEDAKA1.ncrna.fa. If using BLAST on command line, **[BLAST Commandline Manual](https://www.ncbi.nlm.nih.gov/books/NBK279690/)**.

* Build BLAST index:


        makeblastdb -in Oryzias_latipes.ASM223467v1.ncrna.fa -input_type fasta -dbtype nucl -title MEDAKA.ncrna_db -out MEDAKA.ncrna_db >make_blast.txt

* BLAST transcript to ncRNA of related species:

        blastn -evalue 10e-10 -db MEDAKA.ncrna_db -query trinity_out_dir/Trinity.bestisoform.fa -out trinity_bestisoform_fasta_file_MEDAKAncrna.blastout -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq" >blast.txt #blast.txt will be an empty file

Custom scripts are used to identify top BLAST hit (**06.get-ncrna.py**) and filter the assembly (**07.filter-assembly-ncrna.py**):

        python 06.get-ncrna.py trinity_bestisoform_fasta_file_MEDAKAncrna.blastout >06_id_blast_tophits.txt
Create a new folder and put the output of the previous command (.tophits) into this folder.

        python 07.filter-assembly-ncrna.py picta_blast_tophits/ trinity_out_dir/Trinity.bestisoform.fa >07_filter_assembly.txt

This output will appear in the trinity_out_dir folder.

**[Transdecoder](https://github.com/TransDecoder/TransDecoder/wiki)**: Remove transcripts without open reading frames (ORFs).
* Extract long ORFs:

        TransDecoder/TransDecoder.LongOrfs -t trinity_out_dir/Trinity.bestisoform_ncrnafiltered.fa -m 150 >extract_orfs.txt

* Final coding region predictions:

        TransDecoder/TransDecoder.Predict -t trinity_out_dir/Trinity.bestisoform_ncrnafiltered.fa --retain_long_orfs_length 150 >codingregion_predicts.txt 2>codingregion_predicts_2.txt


### d. SEX-DETector
SEX-DETector requires a gametologs to coassemble into one single transcript. We need to further assemble the transcripts into contigs.

**[CAP3](http://seq.cs.iastate.edu/cap3.html)**: Assemble contigs.

        CAP3/cap3 Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds -s 900 >cap3_assembly.txt
        
* Merge CAP3 singlets and contigs:

        cat Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds.cap.contigs Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cds.cap.singlets >Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa


**[BWA](http://bio-bwa.sourceforge.net/)**: Map trimmed reads to the final assembly
* Build BWA index:

        bwa index Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa >bwa_index.txt
        
* Align reads: Example Command:

        bwa aln -t 12 Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa suriname_picta/Trimmed_Reads/T1-Dame_R1_098.output_forward_paired_50bp.fastq.gz >T1-Dame_forward.sai 2>T1-Dame_forward.txt

* Generate SAM format: Example Command:

        bwa sampe -r "@RG\tID:sample_name" Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa picta_bwa/T1-Dame_forward.sai picta_bwa/T1-Dame_reverse.sai suriname_picta/Trimmed_Reads/T1-Dame_R1_098.output_forward_paired_50bp.fastq.gz suriname_picta/Trimmed_Reads/T1-Dame_R2_098.output_reverse_paired_50bp.fastq.gz >picta_sam/T1-Dame_sampe.sam 2>picta_bwa_txt_outputs/T1-Dame_sampe.txt


**[SAMtools](http://www.htslib.org/doc/samtools.html)**: Convert SAMs to BAMs
* Build index:

        samtools faidx Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa
        
* Compress into BAM files: Example Command:

        samtools view -t Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa.fai -F 4 -h -S -b -o ~/picta_bam/T1-Dame_sampe.bam ~/picta_sam/T1-Dame_sampe.sam
        
* Order individual BAM files: Example Command:

        samtools sort -o ~/picta_bam_sorted/T1-Dame.sorted ~/picta_bam/T1-Dame_sampe.bam

**[reads2snp](https://kimura.univ-montp2.fr/PopPhyl/index.php?section=tools)**: Genotype individuals at each locus. You will need a text-file output that includes all the individual sorted BAM files of one family and the PATHs. Example Command:

        reads2snp -min 3 -par 0 -aeb -bqt 20 -rqt 10 -bamlist T1-Family.txt -bamref Trinity.bestisoform_ncrnafiltered.fa.transdecoder.cap.fa -out T1
        
Note the output files:
* .alr - Tab delimited file that shows each position of each contig on a different line with the majoritary allele (most common allele in number of reads). The site is labelled as "M" for monomorphic or "P" for polymorphic. Each individual's total read number is reported and the number of reads for each base [A/C/G/T] if the position is polymorphic is reported.
* .gen - Tab delimited file showing each position of each contig on a separate line with the position number starting from 1, then for each individual the inferred genotype followed by a pipe and the posterior probability of the inferred genotype.

First prepare SEX-DETector files. This will report a Generate the gen_summary file. Example Command:

        sex-detector/Source/SEX-DETector_prepare_file.pl T1.gen T1.alr_gen_summary -hom T1-F1,T1-F2,T1-F3,T1-F5,T1-F6 -het T1-M3,T1-M5,T1-M6,T1-M7,T1-M8 -hom_par T1-Dame -het_par T1-Sire >T1-sexdet-prep.txt
        
Then Run SEX-DETector. Example Command:

        sex-detector/Source/SEX-DETector.pl -alr_gen T1.gen -alr_gen_summary T1.alr_gen_summary -out ~/picta_SEXDETector/T1 -system xy -hom T1-F1,T1-F2,T1-F3,T1-F5,T1-F6 -het T1-M3,T1-M5,T1-M6,T1-M7,T1-M8 -hom_par T1-Dame -het_par T1-Sire -alr T1.alr -seq -detail -detail-sex-linked >T1-sexdet-run.txt

### e. Detecting False Positives:
If possible, use the reference genome of the organism that you are studying to test if SEX-DETector has accurately called sex-linked genes. The genes that are called as X/Y or X/O sex-linked should map to the sex chromosome of the organism.

        bwa index picta_genome.fasta >picta_genome_bwa_index.txt

SEX-DETector outputs the sequences of the sex-linked genes. Use these sequences to determine if they align to the sex chromosome. Example Command:

        bwa aln picta_genome.fasta picta_project/picta_SEXDETector/T1/T1_sex-linked_sequences.fasta >picta_project/picta_SEXDETector/T1_XY_picta_genome.sai 2>picta_project/picta_SEXDETector/T1-aln.txt
        bwa samse -r "@RG\tID:sample_name"  picta_genome.fasta picta_project/picta_SEXDETector/T1_XY_picta_genome.sai picta_project/picta_SEXDETector/T1/T1_sex-linked_sequences.fasta >T1-aln-samse.sam


### f. Using Male-Biased Gene Expression to Detect Additional Y genes:
It is possible that the divergence between X- and Y-chromosome would make it difficult for SEX-DETector to accurately call XY sex-linked genes. Thus, we used [SALMON](https://combine-lab.github.io/salmon/) to get gene expression counts. Normalization and differential gene expression was done using the R scrip found in [Poecilia_picta_Evol_Hist/Ppicta_SEXDETector_Scripts](https://github.com/ljmfong/Poecilia_picta_Evol_Hist/tree/main/Ppicta_SEXDETector_Scripts)

* Build an Index with SALMON. Example Command:

       /salmon-1.4.0/bin/salmon index -t trinity_transdecoder_reference.transcripts.fa -d salmon_prep.txt -p 12 -i salmon_index

* Count using QUANT:

        /salmon-1.4.0/bin/salmon quant -i salmon_index -l A -1 forward_paired.fq -2 reverse_paired.fq --validateMappings -o sample_quant


