## _P. picta_ Evolutionary Rates
This pipeline finds the dN and dS values for the sex-linked genes called by SEX-DETector. Included are orthologs acquired from **[Ensembl](http://uswest.ensembl.org/index.html)**, included are _Oryzias latipes_, _Xiphophorus maculatus_, _Gambusia holbrooki_, _Poecilia latipinna_, _P. formosa_, _P. reticulata_, and the sex-linked genes of _P. wingei_ from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA591249](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/bioproject/?term=PRJNA591249)** and the sex-linked genes of _P. picta_ as predicted by the SEX-DETector pipeline (using the transdecoder.cap.fa file).
Scripts can be found under [Poecilia_picta_Evol_Hist/Ppicta_Evol_Rate_Scripts/](https://github.com/ljmfong/Poecilia_picta_Evol_Hist/tree/main/Ppicta_Evol_Rate_Scripts).

------------------------------------------------------------------------------------------------------------------------------------
### a. Pairwise Divergence of X/Y sex-linked genes in _P. picta_
You can estimate the evolutionary rates of orthologous sequences using **[PAML](http://abacus.gene.ucl.ac.uk/software/paml.html)**. When running PAML, run the program ONCE at a time, it may have issues when running multiple instances of the program at the same time. For pairwise divergence, Run PAML using yn00 instead of codeml, no need to for masking. 
Each gene needs a folder in which there is a .phy file (with just X- and Y-sequences) and the .ctl file that PAML uses to run yn00. 

Create a directory for each _P. picta_ X/Y sex-linked genes that contains both the X- and Y-sequences. Align the two sequences to each other using **[PRANK](http://wasabiapp.org/software/prank/)**. Run PRANK inside each directory, Example Command:

        prank -d=TRINITY_DN1465_c6_g1_i4.p1.fasta -o=TRINITY_DN1465_c6_g1_i4.p1.prank -DNA -codon -once -f=fasta

Move outputs for all orthologs into one folder, e.g. 'orfs'. Check for and remove gaps:

    python 11.check-gaps_python3.py orfs/ -cutoff 300
    python 11.remove-gaps_python3.py orfs/ orfs_invalid/ -cutoff 300
    python 12.convert-fasta-phylip.py orfs prank.gapsrm.fa 

Put all the individual gene-folders into a home folder, for example 'yn00_XY_genes'. NOTE - PAML does not run well with stop codons at the end of the sequence, remove them from the sequences.

    python 13.make-paml-input_yn00.py yn00_XY_genes/
    python 14.run-paml_yn00.py yn00_XY_genes/ > XY-genes-yn00.txt


### b. Identify reciprocal orthologs
Identify reciprocal orthologs via BLAST from fasta files. Get PEPTIDE sequences *cds*, not cdna). Naming of fasta files must be as follows: Speciesname.fa eg Xiphophorusmaculatus.fa. 

    python 00.get-longest-isoform.py 
    
This script processes an Ensembl fasta file and picks the longest isoform for each gene. Outputs a new fasta file ending in 'longest.fasta'. This script will only work if fasta file is in Ensembl format. You should be in the director of interest when running the above script.

**[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)** all the species against each other and then find the top BLAST hits. Before running the script, also make a folder 'best_isoforms' that contains sequence information for orthologs (e.g. the 'longest.fasta' files).
Example Commands:

    python 01.run-blastall_updated.py -e 10e-10 -b blastn -p 4 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq" ~/best_isoform/Poeciliareticulata_longest.fasta  ~/best_isoform/Poeciliawingei_longest.fasta  ~/best_isoform/Poecilialatipinna_longest.fasta  ~/best_isoform/PoeciliapictaX_longest.fasta  ~/best_isoform/PoeciliapictaY_longest.fasta  ~/best_isoform/Xiphophorusmaculatus_longest.fasta  ~/best_isoform/Gambusiaaffinis_longest.fasta  ~/best_isoform/Poeciliaformosa_longest.fasta ~/best_isoform/Poeciliamexicana_longest.fasta
    python 02.top-blasthit.py blastall/ blastall_tophits
    python 03.ortho-cluster.py blastall_tophits.pkl orthocluster PoeciliapictaX,PoeciliapictaY,Poeciliareticulata,Poeciliawingei,Poecilialatipinna,Poeciliamexicana,Poeciliaformosa,Xiphophorusmaculatus,Gambusiaaffinis 

   
The **03.ortho-cluster.py** script takes the pickle output of reciprocal orthologs and identifies clusters of reciprocal orthologs

### c. Identify Open-Reading Frames (ORFs):
Identify open reading frames of orthologous groups of sequences via BLASTx.  

For **04.make-blastx-input.py**, you must specify one of the species that will form the reference based on which open reading frames will be identified in the other species. The chosen species is usually the one with the best annotated genome and the user must supply a fasta file containing protein sequences from this species. 

    python 04.make-blastx-input.py Xiphophorus_maculatus.X_maculatus-5.0-male.pep.all.fa Xiphophorusmaculatus orthocluster.pkl ~/best_isoform/ blastx/ 
    
You will now have a directory that is filed with the orthologs with sequences, run script **05.run-blastx.py**.

    python 05.run-blastx.py blastx -e 10e-10 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qframe sframe sseq" -b blastx 
    
Identify ORFs using **06.orf-prediction.py** and then get stats of the predicted ORFs (**07.orf-stats.py** and **08.check-protein.py**):

    python 06.orf-prediction.py blastx blastx_invalid/
    python 07.orf-stats.py blastx 150
    python 08.check-protein.py blastx blastx_invalid/ Xiphophorus_maculatus.X_maculatus-5.0-male.pep.all.fa

### d. Align orthologs:
In order to align the orthologs, run PRANK, rooted by the outgroup (_X. maculatus_) and all the organisms of interest (e.g. _X. maculatus_, _G. holbrooki_, _P. latipinna_, _P. formosa_, _P. reticulata_, _P. wingei_ X- and Y-, _P. picta_ X- and Y-). Example Commands:

    prank -d=TRINITY_DN1465_c6_g1_i4.p1.prank.best.fas -t=no_mexicana_tree.txt -o=TRINITY_DN1465_c6_g1_i4.p1.prank.rooted -DNA -codon -once -f=fasta


### e. Creating Maximum Likelihood Trees:
Maximum likelihood trees can be created from PAML outputs using **[RAxML](https://cme.h-its.org/exelixis/web/software/raxml/)**.  Example Command:

    raxmlHPC-PTHREADS -f a -p 12345 -s TRINITY_DN1465_c6_g1_i4.p1.prank.rooted.best.fas -x 12345 -# 100 -m GTRGAMMA -n TRINITY_DN1465_c6_g1_i4.p1 -k -o Xiphophorusmaculatus
