## _P. picta_ Evolutionary Rates
This pipeline finds the dN and dS values for the sex-linked genes called by SEX-DETector. Included are orthologs acquired from **[Ensembl](http://uswest.ensembl.org/index.html)**, included are _Oryzias latipes_, _Xiphophorus maculatus_, _Gambusia holbrooki_, _Poecilia latipinna_, _P. formosa_, _P. reticulata_, and the sex-linked genes of _P. wingei_ from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA591249](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/bioproject/?term=PRJNA591249)** and the sex-linked genes of _P. picta_ as predicted by the SEX-DETector pipeline (using the transdecoder.cap.fa file).

------------------------------------------------------------------------------------------------------------------------------------

### a. Identify reciprocal orthologs
Identify reciprocal orthologs via BLAST from fasta files. Naming of fasta files must be as follows: Speciesname.fa eg Oryziaslaptipes.fa. 

    python 00.get-longest-isoform.py 
    
This script processes an Ensembl fasta file and picks the longest isoform for each gene. Outputs a new fasta file ending in 'longest.fasta'. This script will only work if fasta file is in Ensembl format. You should be in the director of interest when running the above script.

**[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)** all the species against each other and then find the top BLAST hits:
Example Commands:

    python 01.run-blastall_updated.py  -e 10e-10 -b blastn -p 4 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq" picta_dnds/Poeciliapicta_longest.fasta fastx/best_isoforms/Oryziaslatipes_longest.fasta fastx/best_isoforms/Poeciliareticulata_longest.fasta fastx/best_isoforms/Xiphophorusmaculatus_longest.fasta >blastall.txt
    python 02.top-blasthit.py blastall/blastall_tophits >blastall_tophits.txt
    python 03.ortho-cluster.py blastall_tophits.pkl ortho_cluster Poeciliapicta,Poeciliareticulata,Xiphophorusmaculatus,Oryziaslatipes
    
The **03.ortho-cluster.py** script takes the pickle output of reciprocal orthologs and identifies clusters of reciprocal orthologs

### b. Identify Open-Reading Frames (ORFs):
Identify open reading frames of orthologous groups of sequences via BLASTx.  

For **04.make-blastx-input.py**, you must specify one of the species that will form the reference based on which open reading frames will be identified in the other species. The chosen species is usually the one with the best annotated genome and the user must supply a fasta file containing protein sequences from this species. Before running the script, also make a folder 'best_isoforms' that contains sequence information for orthologs (e.g. the 'longest.fasta' files).

    python 04.make-blastx-input.py Oryzias_latipes.MEDAKA1.pep.all.fa Oryziaslatipes ortho_cluster.pkl best_isoforms/ blastx/ >make-blastx.txt
    
You will now have a directory that is filed with the orthologs with sequences, run script **05.run-blastx.py**.

    python 05.run-blastx.py blastx/ -e 10e-10 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qframe sframe sseq" -b blastx >run_blastx.txt

Identify ORFs using **06.orf-prediction.py** and then get stats of the predicted ORFs (**07.orf-stats.py** and **08.check-protein.py**):

    python 06.orf-prediction.py blastx/ blastx_invalid/ >orf_pred.txt
    python 07.orf-stats.py blastx/ 150 >orf_stats.txt
    python 08.check-protein.py blastx blastx_invalid/ Oryzias_latipes.MEDAKA1.pep.all.fa >check_protein.txt

### c. Align orthologs:
In order to align the orthologs, use **[PRANK](http://wasabiapp.org/software/prank/)**. First you must prepare flie to run PRANK:

    python 09.make-prank-input.py blastx stripped.cdna.fa >prank_prep.txt

This will keep only the '.stripped.cdna.fa' files. You can now run PRANK on these files. PRANK automatically runs with -DNA, -codon and -once, but you must provide a file with a rooted tree structure, example: (((Sp1,Sp2),Sp3),Sp4);. ! Note the ";" at the end of the tree. Make this tree file as a 'tree.txt' file. Check for gaps and then remove them.

    python 10.run-prank.py blastx/ tree.txt >prank.txt
    python 11.check-gaps_python3.py blastx -cutoff 300 >check_gaps.txt
    python 11.remove-gaps_python3.py blastx prank_invalid/ -cutoff 300 >remove_gaps.txt
    

### d. PAML
    




