## _P. picta_ Evolutionary Rates
This pipeline finds the dN and dS values for the sex-linked genes called by SEX-DETector. Included are orthologs acquired from **[Ensembl](http://uswest.ensembl.org/index.html)**, included are _Oryzias latipes_, _Xiphophorus maculatus_, _Gambusia holbrooki_, _Poecilia latipinna_, _P. formosa_, _P. reticulata_, and the sex-linked genes of _P. wingei_ from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA591249](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/bioproject/?term=PRJNA591249)** and the sex-linked genes of _P. picta_ as predicted by the SEX-DETector pipeline (using the transdecoder.cap.fa file).

------------------------------------------------------------------------------------------------------------------------------------

### a. Identify reciprocal orthologs
Identify reciprocal orthologs via BLAST from fasta files. Naming of fasta files must be as follows: Speciesname.fa eg Oryziaslaptipes.fa. 

    python 00.get-longest-isoform.py 
    
This script processes an Ensembl fasta file and picks the longest isoform for each gene. Outputs a new fasta file ending in 'longest.fasta'. This script will only work if fasta file is in Ensembl format. You should be in the director of interest when running the above script.

**[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)** all the species against each other and then find the top BLAST hits:
Example Commands:

    python 01.run-blastall_updated.py  -e 10e-10 -b blastn -p 4 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq" picta_dnds/Poeciliapicta_longest.fasta fastx/best_isoforms/Oryziaslatipes_longest.fasta fastx/best_isoforms/Poeciliareticulata_longest.fasta fastx/best_isoforms/Xiphophorusmaculatus_longest.fasta >blastall.txt
    python 02.top-blasthit.py blastall/blastall_tophits >blastall_tophits.txt
    python 03.ortho-cluster.py blastall_tophits.pkl ortho_cluster Poeciliapicta,Poeciliareticulata,Xiphophorusmaculatus,Oryziaslatipes
    
The **03.ortho-cluster.py** script takes the pickle output of reciprocal orthologs and identifies clusters of reciprocal orthologs

### b. Identify Open-Reading Frames (ORFs):
Identify open reading frames of orthologous groups of sequences via BLASTx.  

For **04.make-blastx-input.py**, you must specify one of the species that will form the reference based on which open reading frames will be identified in the other species. The chosen species is usually the one with the best annotated genome and the user must supply a fasta file containing protein sequences from this species. Before running the script, also make a folder 'best_isoforms' that contains sequence information for orthologs (e.g. the 'longest.fasta' files).

    python 04.make-blastx-input.py Oryzias_latipes.MEDAKA1.pep.all.fa Oryziaslatipes ortho_cluster.pkl best_isoforms/ blastx/ >make-blastx.txt
    
You will now have a directory that is filed with the orthologs with sequences, run script **05.run-blastx.py**.

    python 05.run-blastx.py blastx/ -e 10e-10 -f "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qframe sframe sseq" -b blastx >run_blastx.txt

Identify ORFs using **06.orf-prediction.py** and then get stats of the predicted ORFs (**07.orf-stats.py** and **08.check-protein.py**):

    python 06.orf-prediction.py blastx/ blastx_invalid/ >orf_pred.txt
    python 07.orf-stats.py blastx/ 150 >orf_stats.txt
    python 08.check-protein.py blastx blastx_invalid/ Oryzias_latipes.MEDAKA1.pep.all.fa >check_protein.txt

### c. Align orthologs:
In order to align the orthologs, use **[PRANK](http://wasabiapp.org/software/prank/)**. First you must prepare flie to run PRANK:

    python 09.make-prank-input.py blastx stripped.cdna.fa >prank_prep.txt

This will keep only the '.stripped.cdna.fa' files. You can now run PRANK on these files. PRANK automatically runs with -DNA, -codon and -once, but you must provide a file with a rooted tree structure, example: (((Sp1,Sp2),Sp3),Sp4);. ! Note the ";" at the end of the tree. Make this tree file as a 'tree.txt' file. Check for gaps and then remove them.

    python 10.run-prank.py blastx/ tree.txt >prank.txt
    python 11.check-gaps_python3.py blastx -cutoff 300 >check_gaps.txt
    python 11.remove-gaps_python3.py blastx prank_invalid/ -cutoff 300 >remove_gaps.txt
    
Your final files will end with '.stripped.prank.gapsrm.fa'

### d. PAML:
You can estimate the evolutionary rates of orthologous sequences using **[PAML](http://abacus.gene.ucl.ac.uk/software/paml.html)**. When running PAML, run the program ONCE at a time, it may have issues when running multiple instances of the program at the same time.

    python 12.convert-fasta-phylip.py blastx stripped.prank.gapsrm.fa >prank-to-phylip.txt
    
Check to make sure that the number of folders remaining in the blastx folder match the numbes you see in the check_gaps output. Make an unrooted tree text file, the file structure will look like: ((Sp1,Sp2),Sp3,Sp4);. ! Note the ";" at the end of the tree.

    python 13.make-paml-input-cleandataoff.py blastx unrooted_tree.txt -model 0 -nssites 0 -fixomega 0 -omega .4 >make-paml-input.txt
    python 14.run-paml.py blastx >run-paml.txt

Rename the blastx folder to 'paml0' to know that this PAML was used to run the branch model 0 (Calculates one omega for the **whole tree**, unrooted tree)

### e. SWAMP:
**[SWAMP](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4251194/)** is a program that masks poorly aligned reads specific for PAML. For example, you may want to run SWAMP firs to mask long stretches of nonhomologous sequences cause by splicing or misannotation, and then run SWAMP again to mask short stretches of sequencing errors, both of which can bias estimates of positive selection. 

    python SWAMP_bracket.py -i ~/picta_dnds/paml0/ -b ~/picta_dnds/branchcode.txt -m 100 -t 6 -w 15 >~/picta_dnds/txt_outputs/swamp-1.txt
    python 0.clean-folder-rename.py paml0/ masked.phy .phy
    python SWAMP_bracket.py -i paml0/ -b branchcode_picta.txt -m 100 -t 2 -w 5  >swamp-2.txt
    python 0.clean-folder.py paml0/ __masked.phy
    
Check for any errors and remove any errors:

    python 15.check-N_python3.py paml0/ Oryziaslatipes,Poeciliareticulata,Poeciliapicta,Xiphophorusmaculatus
    python 15.remove-N_python3.py  paml0/ paml0_swamp_invalid/ Oryziaslatipes,Poeciliareticulata,Poeciliapicta,Xiphophorusmaculatus -cutoff 300 >remove_Ns.txt
    python 12.convert-fasta-phylip.py  paml0/ __masked.phy >keep_only_masked.phy

### f. Running PAML on other Models:
There are different models that can be run with PAML. For example, the previous model (model 0) calculates the evolutionary rate for the whole tree (nssites 0), another model allows you to look at branch-site evolutionary rates (model 2), allowing for omega to vary both among lineages and among sites. You can also test for positive selection by using a "#" in front of the species of interest in your tree text file, for example: ((Sp1#1,Sp2),Sp3,Sp4);
You can test for a null model (labelled below as 'pict_null') and an alternative model (labelled below as 'pict_alt').

* Branch-Site Model:

    python 13.make-paml-input.py pict_null/ unrooted_tree_hash.txt -model 2 -nssites 2 -fixomega 1 -omega 1
    python 14.run-paml.pypict_null/ >paml-run-null-branch-site-model.txt
    python 13.make-paml-input.py pict_alt/ unrooted_tree_hash.txt -model 2 -nssites 2 -fixomega 0 -omega .4
    python 14.run-paml.py pict_alt/ >paml-run-alt-branch-site-model.txt

With the null and alternative models, you can extract log likelihood differences between the two models: -2DL = -2*(likelihood_null-likelihood_alternative). PAML has a function called Chi2 which takes the -2DL and degrees of freedom and outputs the p-value. Degrees of freedom are calculated as the difference in parameters between the null and alternative models. df = np_null - np_alternative.

    python 16.paml-extract-loglikelihoods.py pict_null pict_alt/ loglikelihood.txt
    
 * Branch Model where The Branch Model 0 = pamlmod0; Branch Model 2 = paml2:

    python 13.make-paml-input.py pamlmod0 unrooted_tree.txt -model 0 -nssites 0 -fixomega 0 -omega .4
    python 14.run-paml.py pamlmod0 >paml-branch-model-0.txt
    python 13.make-paml-input.py paml2 unrooted_tree_hash.txt -model 2 -nssites 0 -fixomega 0 -omega .4
    python 14.run-paml.py paml2 >paml-branch-model-2.txt

### g. Comparing Models:
* Model 1:
    python 16.paml-extract-branch-lengths-dSfilter.py pamlmod0/ 6..3 model0-dSfilter.txt
    python 16.paml-extract-branch-lengths-dSSdSfilter.py pamlmod0/ 6..3 model0-dSSdSfilter.txt
    
* Model 2:

    python 16.paml-extract-branch-lengths-dSfilter.py paml2 6..3 model2-dSfilter.txt
    python 16.paml-extract-branch-lengths-dSSdSfilter.py paml2 6..3 model2-dSSdSfilter.txt
    
* Randomization: Since the _P. picta_ sequences used include the X- and Y-sequences inferred from SEX-DETector, you will also need to find the position of where these genes map to the _P. picta_ genome (this has bee done in the SEX-DETector pipeline). Compare the values for X/Y, X/0, all sex-linked, and autosomal genes in _P. picta_. 

    python 19.randomisation_dnds.py Ppic_XY_position.txt XY_dN_bootstrap XY_dS_bootstrap XY_dN_dS_bootstrap
    python 19.randomisation_dnds.py Ppic_XO_position.txt XO_dN_bootstrap XO_dS_bootstrap XO_dN_dS_bootstrap
    python 19.randomisation_dnds.py Ppic_sexlinked.txt sexlinked_dN_bootstrap sexlinked_dS_bootstrap sexlinked_dN_dS_bootstrap
    python 19.randomisation_dnds.py Ppic_autosome.txt autosome_dN_bootstrap autosome_dS_bootstrap autosome_dN_dS_bootstrap    

The output of this file will include 'dN_list lower quartile', 'dN_list upper quartile', 'median dN_list', 'dS_list lower quartile', 'dS_list upper quartile', 'median dS_list', 'dN_dS_list lower quartile', 'dN_dS_list upper quartile', and 'median dN_dS_list'.

* Find p-values:

    python 20.dnds_permutations.py Ppic_autosome.txt Ppic_sexlinked.txt
    python 20.dnds_permutations.py Ppic_autosome.txt Ppic_XY_position.txt
    python 20.dnds_permutations.py Ppic_autosome.txt Ppic_XO_position.txt
    
Output of these files include 'check dN_b, dS_b, dNdS_b', 'check dN_s, dS_s, dNdS_s', 'Number of times no replacement', 'Number of times with replacement', 'dS_p', 'dN_p', 'dNdS_p'.

### h. Finding Pairwise Divergence of X/Y sex-linked genes in _P. picta_
