## Characterizing the Y Chromosome of _P. picta_

This pipeline identifies the covergae differences between female and male _P. picta_. Following the coverage analysis of _P. picta_ paired-end DNA-seq, the contigs from **[Charlesworth et al. (2021) GBE](https://doi-org.ezproxy.library.ubc.ca/10.1093/gbe/evab171)** were characterized using coverage analysis. Scripts used for this analysis can be found under Main/Characterizing_Y_Chromosome_Scripts/

------------------------------------------------------------------------------------------------------------------------------------

### a. Alignment of DNA-seq

Paired-end DNA-seq of _P. picta_ were acquired from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA528814](https://www.ncbi.nlm.nih.gov/sra/PRJNA528814)**.

**[BWA](http://bio-bwa.sourceforge.net/)**: Map reads to the reference genome.
* Build BWA index:

        bwa index Poecilia_picta_genome.fasta >buil_full_picta_index.txt
        
* Align reads (both forward and reverse reads). Example Command:

        bwa aln Poecilia_picta_genome.fasta 247_1.out.fastq.gz >247_autosomes_forward.sai
        bwa aln Poecilia_picta_genome.fasta 247_2.out.fastq.gz >247_autosomes_reverse.sai

* Generate SAM format: Example Command:


      bwa sampe -r "@RG\tID:sample_name" Poecilia_picta_genome.fasta 247_autosomes_forward.sai 247_autosomes_reverse.sai 247_1.out.fastq.gz 247_2.out.fastq.gz >247_autosomes_sampe.sam

    
* Find unique reads: Example Command:

      grep 'XT:A:U' 247_autosomes_sampe.sam >247_autosome_uniq.sam
      
* Find coverage for each individual. Example Command:

      bin/soap.coverage -sam -cvg -refsingle Poecilia_picta_genome.fasta -i 247_autosome_uniq.sam -o 247_autosome_soapCov.txt -p 12 -window 247_autosome_windows.txt 50000

To ensure only autosomes are accounted for - remove the sex chromosome from the soapCov.txt file (in this, Chromosome 12). You can do this manually.
Move the soapcov.txt files into their respective folders (by sex). For example, make a folder labelled "Female" and move all female individuals soapCov.txt files into this directory. Extract the coverage for the sexes. Example Command:

      python bin/06.extract_coverage.py Female/ female_cov.txt
      python bin/06.extract_coverage.py Male/ male_cov.txt
     
Calculate the fold change between the sexes:

      python bin/07.foldchange_coverage.py female_cov.txt male_cov.txt MF_foldchange.txt

      
      




