## Characterizing the Y Chromosome of _P. picta_

This pipeline identifies the covergae differences between female and male _P. picta_. Following the coverage analysis of _P. picta_ paired-end DNA-seq, the contigs from **[Charlesworth et al. (2021) GBE](https://doi-org.ezproxy.library.ubc.ca/10.1093/gbe/evab171)** were characterized using coverage analysis. The contigs were acquired from European Nucleotide Archive under accession number **[ERZ1744533](https://www.ebi.ac.uk/ena/browser/text-search?query=PRJEB43222)**. Scripts used for this analysis can be found under Main/Characterizing_Y_Chromosome_Scripts/

------------------------------------------------------------------------------------------------------------------------------------

### a. Alignment of DNA-seq

Paired-end DNA-seq of _P. picta_ were acquired from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA528814](https://www.ncbi.nlm.nih.gov/sra/PRJNA528814)**.

**[BWA](http://bio-bwa.sourceforge.net/)**: Map reads to the reference genome.
* Build BWA index:

        bwa index Poecilia_picta_genome.fasta >buil_full_picta_index.txt
        
* Align reads (both forward and reverse reads). Example Command:

        bwa aln Poecilia_picta_genome.fasta 247_1.out.fastq.gz >247_autosomes_forward.sai
        bwa aln Poecilia_picta_genome.fasta 247_2.out.fastq.gz >247_autosomes_reverse.sai

* Generate SAM format: Example Command:


      bwa sampe -r "@RG\tID:sample_name" Poecilia_picta_genome.fasta 247_autosomes_forward.sai 247_autosomes_reverse.sai 247_1.out.fastq.gz 247_2.out.fastq.gz >247_autosomes_sampe.sam

    
* Find unique reads: Example Command:

      grep 'XT:A:U' 247_autosomes_sampe.sam >247_autosome_uniq.sam
      
### b. Estimating Coverage

Find coverage for each individual. Example Command:

      bin/soap.coverage -sam -cvg -refsingle Poecilia_picta_genome.fasta -i 247_autosome_uniq.sam -o 247_autosome_soapCov.txt -p 12 -window 247_autosome_windows.txt 50000

To ensure only autosomes are accounted for - remove the sex chromosome from the soapCov.txt file (in this, Chromosome 12). You can do this manually.
Move the soapcov.txt files into their respective folders (by sex). For example, make a folder labelled "Female" and move all female individuals soapCov.txt files into this directory. Extract the coverage for the sexes. Example Command:

      python bin/06.extract_coverage.py Female/ female_cov.txt
      python bin/06.extract_coverage.py Male/ male_cov.txt
     
Calculate the fold change between the sexes:

      python bin/07.foldchange_coverage.py female_cov.txt male_cov.txt MF_foldchange.txt

------------------------------------------------------------------------------------------------------------------------------------

### c. Alignment of _P. picta_ contigs
This maps the _P. picta_ paired-end DNA-seq reads to the contigs that were sequenced from Charlesworth et al. (2021) GBE.
      
**[BWA](http://bio-bwa.sourceforge.net/)**: Map reads to the reference genome.
* Build BWA index:

        bwa index mpicta_GCA_905332405.1.fasta >build-deb-index.txt
        
* Align reads (both forward and reverse reads). Example Command:

        bwa aln mpicta_GCA_905332405.1.fasta 247_1.out.fastq.gz >247_contig_forward.sai
        bwa aln mpicta_GCA_905332405.1.fasta 247_2.out.fastq.gz >247_contig_reverse.sai

* Generate SAM format: Example Command:


      bwa sampe -r "@RG\tID:sample_name" mpicta_GCA_905332405.1.fasta 247_contig_forward.sai 247_contig_reverse.sai 247_1.out.fastq.gz 247_2.out.fastq.gz >247_contig_sampe.sam 2>247-sai-sampe.txt

  
### d. Find coverage of the contigs

Convert the SAM files to BAM files and then order the BAM files using **[SAMtools](http://www.htslib.org/doc/samtools.html)**. Example Command:

        samtools view -S -b 247_contig_sampe.sam >247_contig.bam
        samtools sort -o 247_contig.sorted 247_contig.bam

Use SAMtools to find coverage. Example Command:

        samtools index iulia_picta_dna/deb_coverage/247_contig.sorted
        samtools idxstats iulia_picta_dna/deb_coverage/247_contig.sorted > 247_contigs.idxstats

The output of the '.idxstats' files have 4 Columns. The columsn are: 1) Reference sequence name, 2)Sequence length, 3) Number of mapped read-segments, 4) Number of unmapped read-segments. The reads were normalized by dividing the contig reads by the total number of reads per individual. For example:

        Contig            individiual    sequence_length  number_mapped  number_unmapped total_number_reads      normalized_reads
        CAJOCC010000001.1 female_247    4208005          2169872        10714           387298116               0.00560

To calculate the F:M Read Depth, the median normalized read depth ('normalized_reads') for each sex was calulcated. Bootstrapping and log2 F:M Read Depth was caluclated using the Excel functions INDEX(range, RANDBETWEEN(1, ROWS(range)),1) and IMLOG2, respectively.

### e. Mapping position of contigs
 You will need the organism's reference genome. Use **[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)** to map the position of the contigs.
 * Build BLAST index:


        makeblastdb -in Poecilia_picta_genome.fasta -input_type fasta -dbtype nucl -title Poecilia_picta.db -out Poecilia_picta.db >make_blast.txt
 
 * BLAST the contigs to the genome:


 
        blastn -evalue 10e-10 -db Poecilia_picta.db -query mpicta_GCA_905332405.1.fasta -out contig_positions.out -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq"

* Find the BLAST tophit:

        python 06.get-blast-tophit.py contig_positions.out

