## Characterizing the Y Chromosome of _P. picta_

This pipeline identifies the covergae differences between female and male _P. picta_. Following the coverage analysis of _P. picta_ paired-end DNA-seq, the contigs from **[Charlesworth et al. (2021) GBE](https://doi-org.ezproxy.library.ubc.ca/10.1093/gbe/evab171)** were characterized using coverage analysis. The contigs were acquired from European Nucleotide Archive under accession number **[ERZ1744533](https://www.ebi.ac.uk/ena/browser/text-search?query=PRJEB43222)**. Following the characterization of the contigs, we analyzed the synteny between _P. picta_ and _P. reticulata_. Scripts used for this analysis can be found under [Poecilia_picta_Evol_Hist/Characterizing_Y_Chromosome_Scripts/](https://github.com/ljmfong/Poecilia_picta_Evol_Hist/tree/main/Characterizing_Y_Chromosome_Scripts)

------------------------------------------------------------------------------------------------------------------------------------

### a. Alignment of DNA-seq

Paired-end DNA-seq of _P. picta_ were acquired from **[Darolti et al. (2019) PNAS](https://doi-org.ezproxy.library.ubc.ca/10.1073/pnas.1905298116)** under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA528814](https://www.ncbi.nlm.nih.gov/sra/PRJNA528814)**.

**[BWA](http://bio-bwa.sourceforge.net/)**: Map reads to the reference genome.
* Build BWA index:

        bwa index Poecilia_picta_genome.fasta >buil_full_picta_index.txt
        
* Align reads (both forward and reverse reads). Example Command:

        bwa aln Poecilia_picta_genome.fasta 247_1.out.fastq.gz >247_autosomes_forward.sai
        bwa aln Poecilia_picta_genome.fasta 247_2.out.fastq.gz >247_autosomes_reverse.sai

* Generate SAM format: Example Command:


      bwa sampe -r "@RG\tID:sample_name" Poecilia_picta_genome.fasta 247_autosomes_forward.sai 247_autosomes_reverse.sai 247_1.out.fastq.gz 247_2.out.fastq.gz >247_autosomes_sampe.sam

    
* Find unique reads: Example Command:

      grep 'XT:A:U' 247_autosomes_sampe.sam >247_autosome_uniq.sam
      
### b. Estimating Coverage

Find coverage for each individual. Example Command:

      bin/soap.coverage -sam -cvg -refsingle Poecilia_picta_genome.fasta -i 247_autosome_uniq.sam -o 247_autosome_soapCov.txt -p 12 -window 247_autosome_windows.txt 50000

To ensure only autosomes are accounted for - remove the sex chromosome from the soapCov.txt file (in this, Chromosome 12). You can do this manually.
Move the soapcov.txt files into their respective folders (by sex). For example, make a folder labelled "Female" and move all female individuals soapCov.txt files into this directory. Extract the coverage for the sexes. Example Command:

      python bin/06.extract_coverage.py Female/ female_cov.txt
      python bin/06.extract_coverage.py Male/ male_cov.txt
     
Calculate the fold change between the sexes:

      python bin/07.foldchange_coverage.py female_cov.txt male_cov.txt MF_foldchange.txt

------------------------------------------------------------------------------------------------------------------------------------

### c. Alignment of _P. picta_ contigs
This maps the _P. picta_ paired-end DNA-seq reads to the contigs that were sequenced from Charlesworth et al. (2021) GBE.
      
**[BWA](http://bio-bwa.sourceforge.net/)**: Map reads to the reference genome.
* Build BWA index:

        bwa index mpicta_GCA_905332405.1.fasta >build-deb-index.txt
        
* Align reads (both forward and reverse reads). Example Command:

        bwa aln mpicta_GCA_905332405.1.fasta 247_1.out.fastq.gz >247_contig_forward.sai
        bwa aln mpicta_GCA_905332405.1.fasta 247_2.out.fastq.gz >247_contig_reverse.sai

* Generate SAM format: Example Command:


      bwa sampe -r "@RG\tID:sample_name" mpicta_GCA_905332405.1.fasta 247_contig_forward.sai 247_contig_reverse.sai 247_1.out.fastq.gz 247_2.out.fastq.gz >247_contig_sampe.sam 2>247-sai-sampe.txt

  
### d. Find coverage of the contigs

Convert the SAM files to BAM files and then order the BAM files using **[SAMtools](http://www.htslib.org/doc/samtools.html)**. Example Command:

        samtools view -S -b 247_contig_sampe.sam >247_contig.bam
        samtools sort -o 247_contig.sorted 247_contig.bam

Use SAMtools to find coverage. Example Command:

        samtools index iulia_picta_dna/deb_coverage/247_contig.sorted
        samtools idxstats iulia_picta_dna/deb_coverage/247_contig.sorted > 247_contigs.idxstats

The output of the '.idxstats' files have 4 Columns. The columsn are: 1) Reference sequence name, 2)Sequence length, 3) Number of mapped read-segments, 4) Number of unmapped read-segments. The reads were normalized by dividing the contig reads by the total number of reads per individual. For example:

        Contig            individiual    sequence_length  number_mapped  number_unmapped total_number_reads      normalized_reads
        CAJOCC010000001.1 female_247    4208005          2169872        10714           387298116               0.00560

To calculate the F:M Read Depth, the median normalized read depth ('normalized_reads') for each sex was calulcated. Bootstrapping and log2 F:M Read Depth was caluclated using the Excel functions INDEX(range, RANDBETWEEN(1, ROWS(range)),1) and IMLOG2, respectively.

### e. Mapping position of contigs
 You will need the organism's reference genome. Use **[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)** to map the position of the contigs.
 * Build BLAST index:


        makeblastdb -in Poecilia_picta_genome.fasta -input_type fasta -dbtype nucl -title Poecilia_picta.db -out Poecilia_picta.db >make_blast.txt
 
 * BLAST the contigs to the genome:


 
        blastn -evalue 10e-10 -db Poecilia_picta.db -query mpicta_GCA_905332405.1.fasta -out contig_positions.out -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq"

* Find the BLAST tophit:

        python 06.get-blast-tophit.py contig_positions.out

------------------------------------------------------------------------------------------------------------------------------------

### f. Synteny Analysis
The synteny analysis between _P. picta_ and _P. reticulata_ used the **[MCScanX](https://github.com/wyp1125/MCScanX)** pipeline. We used the _P. picta_ reference genome and the _P. reticulata_ reference genome (we manually corrected for the inversion on the X chromosome). MCScan works best when the organisms of interest are abbreviated to two letters, for example, _P. picta_ is abbreviated to 'pp' and _P. reticulata_ is abbreviated to 'pr'.

We first used BLAST on the peptide sequences for _P. reticulata_ from ENSEMBL (http://ftp.ensembl.org/pub/release-105/fasta/poecilia_reticulata/pep/Poecilia_reticulata.Guppy_female_1.0_MT.pep.all.fa.gz).

 * Build BLAST index:


        makeblastdb -in /Poecilia_reticulata.Guppy_female_1.0_MT.pep.all.fa -dbtype prot -out Poecilia_reticulata.Guppy_female_protein.db
 
 * BLAST the _P. picta_ protein to the _P. reticulata_ proteome:


        blastall -i Poecilia_picta.protein.fasta -d Poecilia_reticulata.Guppy_female_protein.db -p blastp -e 1e-10 -b 5 -v 5 -m 8 -o picta_ensembl_reticulata_protein.blast

This will provide the PROTEIN. In order to get the GENE:
* For _P. reticulata_:

        awk 'BEGIN {FS = OFS = " "}{print $1,$4,$2,$3,$4,$5,$6,$7;}' Poecilia_reticulata.Guppy_female_1.0_MT.pep.all.fa > Poecilia_reticulata.Guppy_female_1.0_MT.pep_genes.fa
        sed -i 's/ENSPREP...................//g' Poecilia_reticulata.Guppy_female_1.0_MT.pep_genes.fa
        
* For _P. picta_:

        awk '{print $1,substr($2,1,18),$3,$4,$5,$6,$7,$8,$9,$10,$11,$12}' picta_ensembl_reticulata_protein.blast >pp.blast

Because MCScan requires collinearity between the two species of interest, you will need to know the positional information for the other species. You will need to create an BLAST index for _P. picta_ as well.

        makeblastdb -in Poecilia_picta.protein.fasta -dbtype prot -out picta_protein.db
        blastall -i Poecilia_reticulata.Guppy_female_1.0_MT.pep_genes.fa -d picta_protein.db -p blastp -e 1e-10 -b 5 -v 5 -m 8 -o pp_1.blast
        awk '{print substr($1,1,18),substr($2,1,8),$3,$4,$5,$6,$7,$8,$9,$10,$11,$12}' pp_1.blast > pr.blast
        cat pp.blast pr.blast >pp_pr.blast
        <pp_pr.blast tr ' ' '\t' > pp_pr_tab.blast
        
Now you will have to get organize the annotation files (.gff). You must ensure that the gene names in the BLAST files created in the previous step match the names found in the GFF files. The .gff file organized for _P. picta_ is as follows:

        awk '{print $1,substr($2,1,8),$3,$4,$5,$6,$7,$8,$9,$10,$11,$12}' pp.blast > pp_test.blast
        grep 'gene' Poecilia_picta.annotation.gff >picta_ensembl_reticulata_protein.gff
        awk 'BEGIN{OFS="\t"}{print $1,$9,$4,$5}' picta_ensembl_reticulata_protein.gff > no_head_picta.gff
        sed -i 's/ID=//g' no_head_picta.gff
        awk '{print $1,substr($2,1,18,$3,$4}' no_head_picta.gff > cleaned_nohead_picta.gff
        
This will replace "Scaffold_" with the species identifier (which was used for the blast) - pp = _P. picta_
        sed -i 's/Scaffold_/pp/g' cleaned_nohead_picta.gff 
        cp cleaned_nohead_picta.gff pp.gff
        
The new 'pp.gff' file will be used for MCScanX. Now the _P. reticulata_ .gff file has to be reorganized to be read by MCScanX.

        awk '{if($3 == "CDS"){print;}}' Poecilia_reticulata.Guppy_female_1.0_MT.105.gtf > Poecilia_reticulata.Guppy_female_1.0_MT.genes.gtf
        awk '{print $1,$28,$4,$5}' Poecilia_reticulata.Guppy_female_1.0_MT.genes.gtf > Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        sed -i 's/"//g' Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        sed -i 's/;//g' Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        awk '{if($3 == "gene"){print;}}' Poecilia_reticulata.Guppy_female_1.0_MT.105.gtf > Poecilia_reticulata.Guppy_female_1.0_MT.genes.gtf
        awk '{print $1,$10,$4,$5}' Poecilia_reticulata.Guppy_female_1.0_MT.genes.gtf > Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        sed -i 's/"//g' Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        sed -i 's/;//g' Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        sed -i 's/LG/pr/g' Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf
        cp Poecilia_reticulata.Guppy_female_1.0_MT.mcscan.gtf pr.gff
 
The new 'pr.gff' file will be used for MCScanX.
        cat pp.gff pr.gff > pp_pr.gff
        
Ensure the '.gff' file is tab-delimited.
        <pp_pr.gff tr ' ' '\t' > pp_pr_tab.gff

Now you can run MCScanX on the files labelled as 'pp_pr'
        MCScanX pp_pr

This will produce a collinearity (synteny) file that you will be able to input into SynVisio to visualize synteny.

### g. Visualizing Synteny
Synteny was visulized using **[SynVisio](https://synvisio.github.io/#/)**.

