## Identifying the Ancestral Sequence in _P. picta_
The pipeline to identify the ancestral sequence in the Poeciliids of interest (_P. picta_, _P. reticulata_, _P. wingei_, and _P. parae_). The sequences used for this analysis are all paired-end DNA-seq. NOTE - this analysis can be computationally extensive. Ensure you have enough RAM and CPU available.
The sequences for _P. picta_ and _P. wingei_ can be found under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA528814](https://www.ncbi.nlm.nih.gov/sra/PRJNA528814)**. The sequences for _P. parae_ can be found under National Center for Biotechnology Information Sequencing Read Archive BioProject ID **[PRJNA714257](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/bioproject/?term=PRJNA714257)**. 
Scripts used can be found under [Poecilia_picta_Evol_Hist/Identifty_Ancestral_Sequence_Scripts/](https://github.com/ljmfong/Poecilia_picta_Evol_Hist/tree/main/Identifty_Ancestral_Sequence_Scripts).

---------------------------------------------------------------------------------------------------------------

### a. K-mer Profile:
_P. picta_ female and male k-mer profiles were first created using **[Jellyfish](https://github.com/gmarcais/Jellyfish)**. NOTE - Jellyfish can only run at one instance at a time. To do so, first pool individuals based on sex.

    cat 247_both.fastq.gz 248_both.fastq.gz 265_both.fastq.gz > all_female_DNA.fastq.gz
    cat 266_both.fastq.gz 267_both.fastq.gz 268_both.fastq.gz > all_male_DNA.fastq.gz

Jellyfish. To make sure forward and reverse sequences are not considered different k-mers, use '-C' to count canonically. There are a number of ways to determine the ideal k-mer size (-m), including balancing between computational time and requirements. Downstream analysis using HAWK will use k-mer sizes of 31bp so we used the same k-mer size for creating the k-mer profile. See HAWK github linked below for more details.

* Count k-mers: Example Command:


      jellyfish-2.3.3 count -C -m 31 -s 100M -t 12 <(zcat all_female_DNA.fastq.gz) 
    
* Create histogram of the counts: Example Command:


      jellyfish-2.3.3 histo ~/picta_kmer/picta_counts/all_female_mer_counts.jf > all_female_kmer_sub_count.histo

* Normalizing histogram based on the number of base pairs. We are normalizing to a target coverage of 30X for 3 individuals (histo acts additively, so you are normalizing to 90X). This can be done by using the equation: (coverage from histo/# of bases) * (genome size * target coverage). # of bases refers to the sum of all base pairs of the individuals. Genome size of _P. picta_ is 743972568bp. Example Command:


      awk '{print (($1/210411314100) * (66957531120),$2}' all_female_DNA_kmer_count.histo  > picta_all_female_kmer_count_norm30x.histo

The k-mer profile is visualized using the histogram_Rscript.

---------------------------------------------------------------------------------------------------------------

### b. HAWK

**[HAWK](https://github.com/atifrahman/HAWK)**, or Hitting Associations with K-mers, is a pipeline that identifies k-mers that are found significantly different from groups of individuals (in this case, we are looking for sex-associated k-mer differences). HAWK must be installed and it will come with its own scripts.HAWK must  be run separately for each species of interest, each species should have its own directory.  We will use _P. picta_ as an example.

In the _P. picta_ directory, copy a version of the 'countKmers' script from the HAWK folder. You will need to edit the script to make sure it finds the correct directory where the reads are located.

    bash countKmers
    
You should have the outputs of 'sorted_files.txt' and 'total_kmer_counts.txt'. Then copy over the 'runHAWK' script from the HAWK folder (will need some modification - see the uploaded scripts used in this pipeline). Ensure that this is in the same directory that has the 'sorted_files.txt' and 'total_kmer_counts.txt'. Ensure the 'sorted_files.txt' has the right pathway to the files outputted by countKmers. Create a 'gwas_info.txt' that looks like this, but without header:

    Individual  Sex   Condition
    247         U     Control
    248         U     Control
    265         U     Control
    266         U     Case
    267         U     Case
    268         U     Case

This will ensure that Sex is not considered as a confounder. 'Control' is now the label for 'Female' and 'Case' is now the label for 'male'.

    bash runHawk
    
NOTE - smartpca ran into an error when running HAWK. Check if runHawk ran without error, but if it has, you must run smartpac separately. Ensure the 'parfile.txt' file in HAWK has the correct number of individuals (e.g. 'numoutevec:  6'). Run smartpca in the folder that has your 'gwas' files.

    HAWK-1.7.0-beta/supplements/EIG6.0.1-Hawk/bin/smartpca -p HAWK-1.7.0-beta/parfile.txt
    
You can now re-run 'runHawk'. Another script was written to skip the first steps of runHawk as they have been completed.

    bash post_runHawk

NOTE - HAWK does not always correctly add up the k-mer counts for the individuals. Therefore, you must ensure that the 'control' and 'case' totals are correct for the files that are outputted from HAWK ('.kmerDiff' files). Columns for control and case files are the same: 

    Kmer-seq	sum_case_kmers	sum_control_kmers	p-value	case_ind_1	case_ind_2	case_ind_3	control_ind_1	control_ind_2	control_ind_3

To get the correct sum k-mers:

    awk '{print $1,$5,$6,$7}' case_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' >sum_case_out_w_bonf_sorted.kmerDiff
    awk '{print $1,$8,$9,$10}' case_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' >sum_case_control_out_w_bonf_sorted.kmerDiff
    cut -f2 sum_case_control_out_w_bonf_sorted.kmerDiff | paste sum_case_out_w_bonf_sorted.kmerDiff - > corrected_sums_case_out_w_bonf_sorted.txt
    
    awk '{print $1,$5,$6,$7}' control_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' > sum_control_case_out_w_bonf_sorted.kmerDiff
    awk '{print $1,$8,$9,$10}' control_out_w_bonf_sorted.kmerDiff | awk '{print $1; for(i=1; i<=NF;i++) j+=$i; print j; j=0 }'  | sed 'N;s/\n/\t/' > sum_control_out_w_bonf_sorted.kmerDiff
    cut -f2 sum_control_out_w_bonf_sorted.kmerDiff | paste ~/picta_HAWK/sum_control_case_out_w_bonf_sorted.kmerDiff - > corrected_sums_control_out_w_bonf_sorted.txt

These files will now have three columns: Kmer-seq	sum_case_counts	sum_control_counts. Run same steps but with different species in their respective folders.

### c. Normalize and Find Sex-Specific k-mers:
Because there are coverage differences between samples, we must normalize the counts from HAWK. In order to do so, we normalized the counts to 20x (25,000,000,000 bp). This was done in each species' folder. Example Command:

    awk -F " " '{print $1,(($2/206050802100)*25000000000),(($3/210411314100)*25000000000)}' ~/picta_HAWK/corrected_sums_case_out_w_bonf_sorted.txt >~/picta_HAWK/test20_normalized_corrected_case_w_bonf.txt
    awk -F " " '{print $1,(($2/206050802100)*25000000000),(($3/210411314100)*25000000000)}' ~/picta_HAWK/corrected_sums_control_out_w_bonf_sorted.txt >~/picta_HAWK/test20_normalized_corrected_control_w_bonf.txt


Now to find the UNIQUE k-mers based on sexes: Example Command:

    awk '{if($3 == "0"){print;}}' ~/picta_HAWK/test20_normalized_corrected_case_w_bonf.txt >~/picta_HAWK/test20_case_normalized_uniq.kmerDiff
    awk '{if($2 == "0"){print;}}' ~/picta_HAWK/test20_normalized_corrected_control_w_bonf.txt >~/picta_HAWK/test20_control_normalized_uniq.kmerDiff

  
We only included k-mers that had a coverage the targeted 20X: Example Command:


    awk -F " " '($2 > 20){print;}' ~/picta_HAWK/test20_case_normalized_uniq.kmerDiff> ~/picta_HAWK/test20_case_normalized_uniq20.kmerDiff
    awk -F " " '($3 > 20){print;}' ~/picta_HAWK/test20_control_normalized_uniq.kmerDiff > ~/picta_HAWK/test20_control_normalized_uniq20.kmerDiff


Use the 'wc -l' command on the final files to find the number of unique k-mers for each species and their respective sex.

### d. Comparing Unique K-mers across species

* Prepare files for comparison: Copy files to a separate directory and name them accordingly for comparison. Example Command:

      cut -f1 -d " " picta_case_normalized_uniq30.kmerDiff > picta_case_normalized_uniq30_cut.kmerDiff
      cut -f1 -d " " picta_control_normalized_uniq30.kmerDiff > picta_control_normalized_uniq30_cut.kmerDiff

Then sort your files: Example Command:

    sort picta_case_normalized_uniq30_cut.kmerDiff > picta_case_normalized_uniq30_cut_sorted.kmerDiff
    sort picta_control_normalized_uniq30_cut.kmerDiff > picta_control_normalized_uniq30_cut_sorted.kmerDiff

 
Compare all species. Use the 'wc -l' on the final comparison files to find the number of shared k-mers. Example Command:



      comm -12 ret_case_normalized_uniq30_cut_sorted.kmerDiff wingei_case_normalized_uniq30_cut_sorted.kmerDiff > shared_RetWin_Male.txt
      comm -12 shared_RetWin_Male.txt parae_case_normalized_uniq20_cut_sorted.kmerDiff > shared_RetWinPar_Male.txt
      
